enhancements/olm/platform-operators.md

```diff
@@ -0,0 +1,435 @@
+  - https://issues.redhat.com/browse/OLM-2570
+  - https://issues.redhat.com/browse/CORS-1873
+  - https://issues.redhat.com/browse/OCPPLAN-7589
+  - https://github.com/openshift/enhancements/blob/master/enhancements/update/cvo-techpreview-manifests.md
+replaces: []
+superseded-by: []
+---
+
+# Platform Operators
+
+## Summary
+
+The enhancement outlines a mechanism that enable support for the lifecycle of platform operators. A platform operator (PO) is an OLM-based cluster extension that can be installed during day 0 operations, and lifecycles with the cluster. This mechanism will ultimately be divided into phases, and this proposal covers the first phase is more detail.
+
+This first phase introduces a basic PO installation mechanism, and helps build the foundational layer for subsequent phases. Future phases may introduce a first-class PO management mechanism for the OLM component.
+
+## Motivation
+
+The motivation for platform operators is two-fold. The strategic direction for OpenShift is to make the core installation smaller by default and at the same time make it more flexible so that customers can tailor their installations to their use cases. This is what we’re calling “Composable OpenShift”.
+
+To make the core smaller, OpenShift is attempting to reduce the CVO-managed payload to the very minimum set of operators required to run the OpenShift control plane. From a control plane perspective, any non-essential operators can be installed after the cluster has been installed (or upgraded). However, from a customer standpoint, their business needs may dictate that these non-essential operators are still included in the lifecycle of cluster installations and upgrades. The Platform Operator (PO) concept will allow cluster administrators to choose which of these non-control-plane-essential operators to include in the cluster lifecycle.
+
+Secondly, some portion of the layered product operators in the OpenShift catalogs are conducive with inclusion in the cluster lifecycle (e.g. cert-manager, service binding operator, and the local storage operator). The Platform Operator concept will open the door for more of OLM’s optional operators to be included in (and impact) the cluster lifecycle. This will allow customers to further tailor their cluster installs and upgrades to their business needs.
+
+### User Stories
+
+- As a cluster admin, I want to be able to install OLM-based Operators at cluster creation, and participate in the cluster's lifecycle.
+- As a cluster admin that installed a PO that's no longer needed, I want to be able to delete a PO without having to uninstall the cluster.
+- As a cluster admin, I want to be able to approve new versions of an individual PO package.
+- As a team that has a direct dependency on an OLM-based Operator, I want to be able to tailor my installation to meet my use case.
+
+### Goals
+
+- Admins can select from a small pool of candidate OLM-based operator packages.
```

Comment:
Right, and this is because we're incrementally rewriting, and introducing, new components of OLM's control plane. The OLM team has been building out this "rukpak" component in the latest 4.x release, which is a new OLM 1.x component that's responsible for applying, and managing, bundle contents. This component doesn't have the same opinionated tenancy model as OLM as it's agnostic to scope, but phase 0 still needs to interface with OLM's legacy registry+v1 bundle format in the interim, which does have scoping implications.

In order to continue using this legacy bundle format, we'll need to restrict which OLM-based operators can be installed to only operators that support being installed a single time per cluster, which may result in a smaller pool of phase 0 candidate operators.

Reply:
but to be clear, we do not ever expect all olm operators to be installable as POs.  POs will always have additional requirements that general olm operators do not have, so the set of POs will always be a subset of "all olm operators"

Reply:
> That's an interesting question. Aside from the desire for a slow roll-out, do we have any particular reason to limit someone from deciding to treat their thing as a platform operator?

not sure if your question is directed at the short term or long term.  In the short term:

* our limited resources to work w/ teams to ensure they do the right things
* the possibility that requirements change and we don't want to break a bunch of people who were trying to adopt this
* the potential for platform operators to break cluster installs/upgrades which, while not our fault, is something we'll end up getting pulled into

in the long term:
No, though we should provide meaningful guidelines about when it is/isn't reasonable to be a platform operator, i don't think we'd have any desire (or ability) to explicitly deny someone from creating one if that's what they want to do.  Unlike restrictions on the payload, the implications of someone creating a new platform operator are far lower risk.  i.e. things in the payload:

increase the payload size (mirroring cost)
can block us from shipping a release
can break installs/upgrades for all cluster users w/ no obvious workaround

by contrast platform operators:
don't affect the payload size
dont block the core OCP release from shipping even if they themselves are broken (unless OCP is breaking them, then maybe)
would only break installs/upgrades for clusters that use that particular PO

Reply:
I was just thinking whether we'd need any processes in place for the intermediate term given these POs can impact the cluster, and I'd like to insulate that the OLM team to avoid that component being the direct interface that receives all of the support cases and bugs when an individual PO fails due to an issue with their operator.

Reply:
Maybe the gate is just allowing that content into the catalog? How is approval for changes to the catalog managed generally?

Reply:
Assuming we're talking in the context of additions to the catalog, I _think_ there's a certification process for the redhat-* catalogs, but I'd defer to @bparees, @kevinrizza, or @joelanford in case they have a definitive answer.

Reply:
> do we have any particular reason to limit someone from deciding to treat their thing as a platform operator?

I don't think so, other than that we'll have an objective criteria to meet the PO bar.

> Maybe the gate is just allowing that content into the catalog?

There are several pieces of the operator pipeline that do operator testing and verification that gate that particular operator version's acceptance into the catalog. But I think we'll need to make those test and verification steps aware of platform operators because (IMO) customers will likely expect them to have the same level of stability and support as the payload.

Do we have a set of criteria for operators that can be accepted into the payload? That might be another good place to look for defining what can be a PO.

Reply:
Do we require any process' in place for this first phase? IIRC, we have code in place that verifies the AllNamespace install modes, and so some of the restrictions in place for this phase should be codified downstream. Given this, does it make sense to mark this as an open question for phase 1?

Reply:
> Do we require any process' in place for this first phase? 

i don't think so, i think we can be pretty loose in saying "here are the things we expect will work, you can try others but they probably won't work and may break your cluster".

Enforcing allnamespace mode is worth having though.

Reply:
Agreed. Just trying to piece together how we should be tracking all of these context/discussions to help shape future phases.